### PR TITLE
DisruptorReceiveChannel fix

### PR DESCRIPTION
**Pull request by SercanKaraoglu**
This is the fix upon #135 please have a look at

### CODE REVIEW

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java [hunk 1]
```diff
@@ -42,18 +43,21 @@ public DisruptorReceiveChannel(RingBuffer<Message> buffer, Sequence... dependent
...
+            while (nextSequence > availableSequence) {
                 availableSequence = barrier.waitFor1(nextSequence);
-            } catch (AlertException ex) {
-                // ???
             }
+            Message message = buffer.get(nextSequence);
+            sequence.set(nextSequence);
+            return message;
+        } catch (AlertException e) {
+            return null;
```
Comment (pron) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java]: What is `AlertException` used for, and why should we return `null` when it is thrown? We can return `null` from `receive` if and only if the channel is closed. If that exception signals that the channel is closed, we must set the closed status.

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java [hunk 2]
```diff
@@ -57,11 +53,13 @@ public long waitFor1(long sequence, Sequence cursorSequence, Sequence dependentS
             try {
                 for (int i = 0; (availableSequence = cursorSequence.get()) < sequence; i++) {
                     barrier.checkAlert();
-                    processorNotifyCondition.await(i);
+                    processorNotifyCondition.await(i, 1, TimeUnit.NANOSECONDS);
```
Comment (pron) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java]: Can't have that. This is a blocking strategy, so we must block.

File: quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java [hunk 3]
```diff
@@ -32,8 +29,9 @@ public SimpleConditionSynchronizer(Object owner) {
     public Object register() {
         final Strand currentStrand = Strand.currentStrand();
         record("register", "%s register %s", this, currentStrand);
-        waiters.add(currentStrand);
-        return null;
+        while (!waiters.offer(currentStrand))
```
Comment (pron) [inline quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java]: Why?

File: quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java [hunk 4]
```diff
@@ -32,8 +29,9 @@ public SimpleConditionSynchronizer(Object owner) {
...
+            ;
+        return currentStrand;
```
Comment (pron) [inline quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java]: Why?

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java (hunk 2 above)
Comment (SercanKaraoglu) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java]: When we start running fiber consumers they never end they always wait for barrier to release for new sequences and If we block the fiber which is a fork join task behind the scene, forkjoinpool doesn't take another fork join task for executing so everything stucking there. Say you have 3 thread running in thread pool with id 1,2,3 In this case they always execute fibers with id 1,2,3 I couldn't find another way to execute other fibers rather than solution up there Please help me on this

Comment (pron) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java]: But when fibers block they don't block the thread they run in (i.e. the FJPool does execute other tasks). Have you tried using Quasar with plain channels? They work the same way. If you haven't I suggest you do, just to make sure everything works as it should.

File: quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java (hunk 3 above)
Comment (SercanKaraoglu) [inline quasar-core/src/main/java/co/paralleluniverse/strands/SimpleConditionSynchronizer.java]: Offer is preferable to add, add can fail to insert

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java (hunk 1 above)
Comment (SercanKaraoglu) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java]: AlertException is thrown when somebody wants to stop the processor by calling halt() of event processors so we may also use it to close the channel

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java (hunk 2 above)
Comment (SercanKaraoglu) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/StrandBlockingWaitStrategy.java]: Okay I fixed it back to processorNotifyCondition.await(i);  commit and pushed the branch again

File: quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java (hunk 1 above)
Comment (SercanKaraoglu) [inline quasar-core/src/main/java/co/paralleluniverse/strands/channels/disruptor/DisruptorReceiveChannel.java]: Okay then let me do another commit for setting the channel close in this case @pron

**Comment by circlespainter** [comment]
@SercanKaraoglu Does this PR (and the related #135, which seems to have been solved) need to stay open?

**Comment by pron** [comment]
No. I have merged this PR manually (in the 0.7.0 branch), and refactored the entire package to a separate project in master.
